src/OneinchCallee.sol

```diff
@@ -0,0 +1,158 @@
+    DaiJoinLike public immutable daiJoin;
+    TokenLike public immutable dai;
+    uint256 public constant RAY = 10**27;
+
+    function _add(uint256 x, uint256 y) internal pure returns (uint256 z) {
+        require((z = x + y) >= x, 'ds-math-add-overflow');
+    }
+
+    function _sub(uint256 x, uint256 y) internal pure returns (uint256 z) {
+        require((z = x - y) <= x, 'ds-math-sub-underflow');
+    }
+
+    function _divup(uint256 x, uint256 y) internal pure returns (uint256 z) {
+        z = _add(x, _sub(y, 1)) / y;
+    }
+
+    constructor(address daiJoin_) public {
+        daiJoin = DaiJoinLike(daiJoin_);
+        TokenLike dai_ = DaiJoinLike(daiJoin_).dai();
+        dai = dai_;
+        dai_.approve(daiJoin_, uint256(-1));
+    }
+
+    function _fromWad(address gemJoin, uint256 wad) internal view returns (uint256 amt) {
+        amt = wad / 10**(_sub(18, GemJoinLike(gemJoin).dec()));
+    }
+
+    function clipperCall(
+        address sender, // Clipper caller, pays back the loan
+        uint256 owe, // Dai amount to pay back          [rad]
+        uint256 slice, // Gem amount received           [wad]
+        bytes calldata data // Extra data, see below
+    ) external {
+        (
+            address to, // address to send remaining DAI to
+            address gemJoin, // gemJoin adapter address
+            uint256 minProfit, // minimum profit in DAI to make [wad]
+            address charterManager, // pass address(0) if no manager
+            address router, // tx.to address received from the 1inch API
+            bytes memory oneinchData // tx.data received from the 1inch API, without first 4 bytes
+        ) = abi.decode(data, (address, address, uint256, address, address, bytes));
+
+        // Convert slice to token precision
+        slice = _fromWad(gemJoin, slice);
+
+        // Exit gem to token
+        if(charterManager != address(0)) {
+            CharterManagerLike(charterManager).exit(gemJoin, address(this), slice);
+        } else {
+            GemJoinLike(gemJoin).exit(address(this), slice);
+        }
+
+        // Approve 1inch to take gem
+        TokenLike gem = GemJoinLike(gemJoin).gem();
+        gem.approve(router, slice);
+
+        // Calculate amount of DAI to Join (as erc20 WAD value)
+        owe = _divup(owe, RAY);
+        uint256 oweWithProfit = _add(owe, minProfit);
+
+        { // Execute 1inch swap
+            (
+                IAggregationExecutor executor,
+                OneinchRouter.SwapDescription memory swapDescription,
+                bytes memory permit,
+                bytes memory tradeData
+            ) = abi.decode(oneinchData, (IAggregationExecutor, OneinchRouter.SwapDescription, bytes, bytes));
+
+            // Overwrite key values
+            swapDescription.amount = slice;
+            swapDescription.minReturnAmount = oweWithProfit;
+
+            // Execute
+            OneinchRouter(router).swap(executor, swapDescription, permit, tradeData);
```

Comment:
```suggestion
            OneinchRouter(router).swap(executor, swapDescription, "", tradeData);
```
See above comment about permit.

Reply:
Why is it better than passing the permit which is `0x` anyway?

Reply:
I think it's cleaner since it removes any doubt to whether that's used and also probably a bit better gas wise.